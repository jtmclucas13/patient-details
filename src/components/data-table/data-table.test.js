import React from 'react';
import { mount, shallow } from 'enzyme';
import DataTable from './data-table';

describe('component DataTable', () => {
    test('renders empty table if no columns and empty data provided', () => {
        const data = [];
        const wrapper = renderDataTable({ data });

        expect(wrapper).toMatchSnapshot();
    });

    test('renders table with proper headers and data', () => {
        const columns = [{ accessor: (data) => data.data, displayName: 'Column 1', key: '123' }];
        const data = [{ data: 'Value' }];

        const wrapper = renderDataTable({ columns, data });

        expect(wrapper).toMatchSnapshot();
    });

    test('renders not found option if no data provided', () => {
        const columns = [{ accessor: (data) => data.data, displayName: 'Column 1', key: '123' }];
        const dataNotFoundMessage = 'Data not found.';

        const wrapper = renderDataTable({ columns }, true);
        const row = wrapper.find('DataTableRow');
        const cellText = row.text();

        expect(cellText).toEqual(dataNotFoundMessage);
    });

    test('applies special class for odd-numbered rows', () => {
        const columns = [{ accessor: (data) => data.data, displayName: 'Column 1', key: '123' }];
        const data = [
            { data: 'Value' },
            { data: 'Other Value' },
        ];
        const numberOfOddRows = data.length / 2;

        const wrapper = renderDataTable({ columns, data });
        const oddRows = wrapper.find('.data-table__row--odd');

        expect(oddRows.length).toEqual(numberOfOddRows);
    });

    test('builds rows from column properties', () => {
        const value = 'Value';
        const columns = [{ accessor: (data) => data.data, displayName: 'Column 1', key: '123' }];
        const data = [{ data: value }];
        const expectedRowText = value;

        const wrapper = renderDataTable({ columns, data });
        const row = wrapper.find('DataTableRow');
        const rowProps = row.props();
        const { cells } = rowProps;
        const cellValue = cells[0].value;

        expect(cellValue).toEqual(expectedRowText);
    });

    test('renders empty cells when data for column is not found', () => {
        const columns = [{ accessor: (data) => data.data, displayName: 'Column 1', key: '123' }];
        const data = [{}];

        const wrapper = renderDataTable({ columns, data });
        const cell = wrapper.find('.data-table__cell');
        const cellChildNodes = cell.children();

        expect(cellChildNodes.length).toEqual(0);
    });
});

/**
* Helper function for rendering DataTable in tests without boilerplate required props
* @param {Object} [props={}] props for DataTable
* @return {ShallowWrapper} enzyme wrapper for DataTable 
*/
function renderDataTable(props = {}, shouldMount = false) {
    const propsToUse = {
        columns: [],
        ...props,
    };

    if (shouldMount) {
        return mount(<DataTable {...propsToUse} />);
    }

    return shallow(<DataTable {...propsToUse} />);
}
