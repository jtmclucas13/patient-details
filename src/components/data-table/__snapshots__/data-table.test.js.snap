// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component DataTable renders empty table if no columns and empty data provided 1`] = `
<table
  className="data-table"
>
  <DataTableHeader
    columns={Array []}
  />
  <tbody
    className="data-table__body"
  />
</table>
`;

exports[`component DataTable renders table with proper headers and data 1`] = `
<table
  className="data-table"
>
  <DataTableHeader
    columns={
      Array [
        Object {
          "accessor": [Function],
          "displayName": "Column 1",
          "key": "123",
        },
      ]
    }
  />
  <tbody
    className="data-table__body"
  >
    <DataTableRow
      cells={
        Array [
          Object {
            "colSpan": undefined,
            "value": "Value",
          },
        ]
      }
      className={null}
      key="0"
      onClick={[Function]}
    />
  </tbody>
</table>
`;
